File: frontend/src/services/user.service.ts
// Unused imports removed for cleanliness.

interface ApiResponse<T = any> {
  status: string;
  data?: T;
  message?: string;
}

class UserService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = `${import.meta.env.VITE_API_URL}/api/users`;
  }

  private getAuthToken(): string | null {
    const TOKEN_KEY = import.meta.env.VITE_AUTH_TOKEN_KEY || 'passport_buddy_token';
    return localStorage.getItem(TOKEN_KEY);
  }

  private async makeRequest<T>(url: string, options: RequestInit = {}): Promise<ApiResponse<T>> {
    const token = this.getAuthToken();
    
    const response = await fetch(url, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...(token && { 'Authorization': `Bearer ${token}` }),
        ...options.headers,
      },
    });

    if (!response.ok) {
        // Use a custom Error class or object if you have one, otherwise this is fine.
        const errorData = await response.json().catch(() => ({ 
            message: 'An unknown error occurred' 
        }));
        const error = new Error(errorData.message || `Request failed with status ${response.status}`);
        // Attach the response object for more context in the calling code
        (error as any).response = response; 
        throw error;
    }

    return response.json();
  }

  async followUser(userId: string): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/follow/${userId}`, {
      method: 'POST',
    });
  }

  async unfollowUser(userId: string): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/follow/${userId}`, {
      method: 'DELETE',
    });
  }

  async getProfile(): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/profile`);
  }

  async getProfileByUsername(username: string): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/profile/${username}`);
  }
  
  async updateAvatar(image: File): Promise<ApiResponse> {
    const formData = new FormData();
    formData.append('avatar', image);

    const token = this.getAuthToken();
    const response = await fetch(`${this.baseUrl}/avatar`, {
        method: 'POST',
        headers: {
            ...(token && { 'Authorization': `Bearer ${token}` }),
        },
        body: formData
    });
    
    if (!response.ok) {
        const errorData = await response.json().catch(() => ({ message: 'Failed to update avatar' }));
        const error = new Error(errorData.message);
        (error as any).response = response;
        throw error;
    }
    
    return response.json();
  }

  async blockUser(userId: string): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/block/${userId}`, {
      method: 'POST',
    });
  }

  async unblockUser(userId: string): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/block/${userId}`, {
      method: 'DELETE',
    });
  }

  // Merged method signature, combining the safety of 'main' with the new field from 'feats'
  async updateProfile(profileData: {
    fullName?: string;
    username?: string; // Added from the 'feats' branch functionality
    bio?: string;
    location?: string;
    homeAirport?: string;
    passportCountry?: string;
  }): Promise<ApiResponse> {
    return this.makeRequest(`${this.baseUrl}/profile`, {
      method: 'PATCH',
      body: JSON.stringify(profileData),
    });
  }
}

export const userService = new UserService();

File: mobile/lib/services/user_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../core/api/api_config.dart';

class UserService {

  static Future<Map<String, dynamic>?> getUserProfile(String userId) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.get(
        Uri.parse('$baseUrl/api/v1/users/$userId'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        print('Failed to load user profile: ${response.statusCode}');
        return null;
      }
    } catch (e) {
      print('Error fetching user profile: $e');
      return null;
    }
  }

  static Future<bool> updateUserProfile(String userId, Map<String, dynamic> updates) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.put(
        Uri.parse('$baseUrl/api/v1/users/$userId'),
        headers: {
          'Content-Type': 'application/json',
        },
        body: json.encode(updates),
      );

      return response.statusCode == 200;
    } catch (e) {
      print('Error updating user profile: $e');
      return false;
    }
  }

  static Future<List<Map<String, dynamic>>> searchUsers(String query) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.get(
        Uri.parse('$baseUrl/api/v1/users/search?q=$query'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        return data.cast<Map<String, dynamic>>();
      } else {
        print('Failed to search users: ${response.statusCode}');
        return [];
      }
    } catch (e) {
      print('Error searching users: $e');
      return [];
    }
  }

  static Future<bool> followUser(String userId) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/v1/users/$userId/follow'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      return response.statusCode == 200;
    } catch (e) {
      print('Error following user: $e');
      return false;
    }
  }

  static Future<bool> unfollowUser(String userId) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.delete(
        Uri.parse('$baseUrl/api/v1/users/$userId/follow'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      return response.statusCode == 200;
    } catch (e) {
      print('Error unfollowing user: $e');
      return false;
    }
  }

  static Future<bool> blockUser(String userId) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.post(
        Uri.parse('$baseUrl/api/v1/users/$userId/block'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      return response.statusCode == 200;
    } catch (e) {
      print('Error blocking user: $e');
      return false;
    }
  }

  static Future<bool> unblockUser(String userId) async {
    try {
      final apiUrl = await ApiConfig.discoverEndpoint();
      final baseUrl = apiUrl.replaceAll('/graphql', '');
      
      final response = await http.delete(
        Uri.parse('$baseUrl/api/v1/users/$userId/block'),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      return response.statusCode == 200;
    } catch (e) {
      print('Error unblocking user: $e');
      return false;
    }
  }
}